src/System.Linq/tests/ZipTests.cs

```diff
@@ -0,0 +1,393 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.Linq.Tests
+{
+    public class ZipTests
+    {
+        [Fact]
+        public void ImplicitTypeParameters()
+        {
+            IEnumerable<int> first = new int[] { 1, 2, 3 };
+            IEnumerable<int> second = new int[] { 2, 5, 9 };
+            IEnumerable<int> expected = new int[] { 3, 7, 12 };
+
+            Assert.Equal(expected, first.Zip(second, (x, y) => x + y));
+        }
+        
+        [Fact]
+        public void ExplicitTypeParameters()
+        {
+            IEnumerable<int> first = new int[] { 1, 2, 3 };
+            IEnumerable<int> second = new int[] { 2, 5, 9 };
+            IEnumerable<int> expected = new int[] { 3, 7, 12 };
+
+            Assert.Equal(expected, first.Zip<int, int, int>(second, (x, y) => x + y));
+        }
+
+        [Fact]
+        public void FirstIsNull()
+        {
+            IEnumerable<int> first = null;
+            IEnumerable<int> second = new int[] { 2, 5, 9 };
+            
+            Assert.Throws<ArgumentNullException>(() => first.Zip<int, int, int>(second, (x, y) => x + y));
+        }
+
+        [Fact]
+        public void SecondIsNull()
+        {
+            IEnumerable<int> first = new int[] { 1, 2, 3 };
+            IEnumerable<int> second = null;
+            
+            Assert.Throws<ArgumentNullException>(() => first.Zip<int, int, int>(second, (x, y) => x + y));
+        }
+        
+        [Fact]
+        public void FuncIsNull()
+        {
+            IEnumerable<int> first = new int[] { 1, 2, 3 };
+            IEnumerable<int> second = new int[] { 2, 4, 6 };
+            Func<int, int, int> func = null;
+            
+            Assert.Throws<ArgumentNullException>(() => first.Zip(second, func));
+        }
+
+        private class MyIEnum<T> : IEnumerable<T>
+        {
+            public IEnumerable<T> _data;
+
+            public MyIEnum(IEnumerable<T> source)
+            {
+                _data = source;
+            }
+
+            public IEnumerator<T> GetEnumerator()
+            {
+                foreach (var datum in _data)
+                {
+                    if (datum.Equals(2)) throw new Exception();
+                    yield return datum;
+                }
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return this.GetEnumerator();
+            }
+        }
+
+        [Fact]
+        public void ExceptionThrownFromFirstGetEnumerator()
```

Comment:
It is indeed, I think I'll rename to `ExceptionThrownFromFirstGetEnumerator`, since the test is fine, just the way I turned the comment on the original into a test name.